### ISSUE TITLE
Bug for calculate a jacobian

### ISSUE THREAD

**Issue by Koishi-love** [SciML/DifferentialEquations.jl#888]
Consider the classical stiff ODE system
![image](https://user-images.githubusercontent.com/58931198/184308328-16c62085-48bb-4a01-8f7f-fcfe30f13c69.png)

It is easy to declare the ODEProblem in julia by following codes with the is-inplace form
`function f(du, u, p, t)'
    'du[1] = -0.04*u[1] + 10.0^4.0 *u[2]*u[3]'
    'du[2] =  0.04*u[1] - 10.0^4.0 *u[2]*u[3] - 3*10.0^7.0*u[2]^2.0'
    'du[3] = 3*10.0^7.0*u[2]^2.0'
'end'
'prob = ODEProblem(f)`

Using RadauIIA5, the equations can be solved efficiently by
`solve(prob, RadauIIA5())`

Now I give the issue.
I define the system in C++, and use "ccall" to call it. The way doesn't matter, it's just like encapsulating the system. There are the codes:
`f = c(du, u, p, t) = ccall((:Robertson_System, "DESystem.dll"), Cvoid, (Ptr{Cdouble}, Ptr{Cdouble}, Cdouble), du, u, t);`
This system, has been defined correctly, and successfully pass to julia.
However, this leads to completely wrong results.
Finally, I found that for the encapsulated system, the Jacobian matrix was completely miscalculated. Because if I pass the correctly jacobian to ODEFunction, desired solutions can be given.

**Comment by ChrisRackauckas** [comment]
Is the Jacobian transposed in the C++ code because it's row major?

**Comment by Koishi-love** [comment]
> Is the Jacobian transposed in the C++ code because it's row major?
In c++, the system is defined in the form
```c++
void Robertson_System(double *du, double *u, double t){
*du = ...
*(du + 1) = ...
*(du + 2) = ...
}
```
Is there anything I should do?

**Comment by ChrisRackauckas** [comment]
You're missing `p`.

**Comment by Koishi-love** [comment]
> 



> Did you double check the values are the same?


They are same, almostly. In fact, by testing 100,000 random vectors, there are more than 200 points that are different, but the error is negligible (1e-9). So that is not the point.

**Comment by ChrisRackauckas** [comment]
Are they the same for `solve(prob, RadauIIA5(autodiff=false))`? I'm surprised you're not getting autodiff errors.

**Comment by Koishi-love** [comment]
> Are they the same for `solve(prob, RadauIIA5(autodiff=false))`? I'm surprised you're not getting autodiff errors.


Yes, now they are same as the desired solutions. Thanks very much. 
But I'm not sure I understand why 'autodiff=false' needs to be set here. I would like to know when this setting is needed, can julia automatically figure it out for me? Otherwise under what situations should I set this option. Could you, please, tell me?

**Comment by ChrisRackauckas** [comment]
Can you share an example I can run? Any attempts to recreate this just throw the appropriate error that C code could not be autodiff'd. 

> But I'm not sure I understand why 'autodiff=false' needs to be set here.

C code cannot be autodiff'd by ForwardDiff.

**Comment by Koishi-love** [comment]
Thanks!
However, I'm sorry that my source files are encrypted. I'm not sure if the '.dll' file sent to you is available.
The file 'DESystem.dll' will be sent to you by email. 
But the system in C is constructed by codes:
```
void Robertson_System(double* du, double* u, double t)
{
	*du = -0.04 * (*u) + pow(10.0, 4.0) * (*(u + 1)) * (*(u + 2));
	*(du + 1) = 0.04 * (*u) - pow(10.0, 4.0) * (*(u + 1)) * (*(u + 2)) - 3 * pow(10.0, 7.0) * pow((*(u + 1)), 2.0);
	*(du + 2) = 3 * pow(10.0, 7.0) * pow((*(u + 1)), 2.0);
}
```

Then you can wrap it in a ’.dll’ file and call it via ’ccall’.
Then RadauIIA3(autodiff = false) will give an error.

**Comment by ChrisRackauckas** [comment]
It's autodiff that should error?

But I see:

```julia
f = c(du, u, p, t) = ccall((:Robertson_System, "DESystem.dll"), Cvoid, (Ptr{Cdouble}, Ptr{Cdouble}, Cdouble), du, u, t);
```

that's some weird syntax right there, does that not just error at the definition? It should be:

```julia
f(du, u, p, t) = ccall((:Robertson_System, "DESystem.dll"), Cvoid, (Ptr{Cdouble}, Ptr{Cdouble}, Cdouble), du, u, t);
```

and that should error?

**Comment by Koishi-love** [comment]
If that definition leads to an error, I shouldn't be able to calculate the correct result no matter what method used.
However, RadauIIA5 can give a desired solution.

**Comment by ChrisRackauckas-Claude** [comment]
Retested during a triage sweep. The reporter's `.dll` was encrypted so a runnable example never arrived here — I compiled the C source posted in the thread with gcc and reproduced it.

```
julia RHS, RadauIIA5() autodiff        -> Success   u[end]=[0.01787, 7.27e-8, 0.98213]
C RHS,     RadauIIA5() autodiff        -> MaxIters  u[end]=[1.0005, 3.63e-5, 0.00129]   <- WRONG
C RHS,     RadauIIA5(AutoFiniteDiff()) -> Success   u[end]=[0.01787, 7.27e-8, 0.98213]  <- matches
```

Mechanism confirmed directly, and it is worse than assumed in this thread: `Vector{Dual}` is a bitstype array, so `ccall(..., Ptr{Cdouble}, ...)` **silently reinterprets Dual memory as doubles**. There is no conversion error to raise — Julia cannot detect this, so the expectation that it "throws the appropriate error that C code could not be autodiff'd" is not achievable at the ccall boundary.

Partial good news: the reporter's last specific complaint is fixed — `RadauIIA3(autodiff=AutoFiniteDiff())` now returns `Success, u[end]=[0.017869, 7.28e-8, 0.98218]` rather than erroring. And the autodiff case now at least reports `retcode=MaxIters` instead of silently succeeding.

Realistic remaining ask is probably documentation plus a `FunctionWrappersWrapper`-style guard, not detection at the ccall.

Verified on DifferentialEquations v8.0.2 / OrdinaryDiffEq v7.1.3 / Julia 1.11.9.

**Comment by ChrisRackauckas-Claude** [comment]
## Docs fix for the ccall + Dual silent-reinterpret failure mode

Julia cannot detect that `ccall(..., Ptr{Cdouble}, ...)` is reading `Vector{Dual}` memory as doubles — there is no conversion error at the ABI. Detection is not a realistic remaining ask; documentation + user/wrapper guards are.

Draft FAQ PR:

- SciML/DiffEqDocs.jl#880

covers:

1. `autodiff = ADTypes.AutoFiniteDiff()` for foreign RHS
2. analytical `jac`
3. guard that rejects non-`AbstractFloat` eltypes before ccall (or `FunctionWrapperSpecialize` so Dual misses with `NoFunctionWrapperFoundError`)
4. explicit note that auto-detection at the C ABI is impossible

`FunctionWrapperSpecialize` already acts as a library-level guard for Dual (wrapper variants must match); raw `ccall` without a Dual-rejecting wrapper remains the dangerous pattern the FAQ warns about.

**Comment by ChrisRackauckas** [comment]
Okay this is understood now, and it's a Base Julia behavior. Taking it up there.

**Comment by ChrisRackauckas-Claude** [comment]
## Base Julia context (not SciML-specific)

The silent Dual-as-`Cdouble` failure mode is a special case of a **general Julia FFI footgun**: arrays passed to `ccall` as `Ptr{S}` are **reinterpreted regardless of element type**. There is no conversion check at the ABI boundary.

Relevant Base issues (both still open):

- [JuliaLang/julia#29850](https://github.com/JuliaLang/julia/issues/29850) — *Arrays passed to ccall by pointer are reinterpreted regardless of the element type* (docs + ffi). Notes that the C-interop manual currently claims arrays are *not* reinterpret-cast and that eltypes must match, which does not match runtime behavior.
- [JuliaLang/julia#60268](https://github.com/JuliaLang/julia/issues/60268) — *unsafe convert in `ccall` when convert Vec{Int64} to Ptr{Int32}* — asks for runtime/compile-time checking when array eltype and `Ptr` target disagree.

### Mechanism (same path Dual hits)

In `base/pointer.jl`, roughly:

```julia
cconvert(::Type{<:Ptr}, a::Array) = getfield(a, :ref)
unsafe_convert(::Type{Ptr{S}}, a::AbstractArray{T}) where {S,T} =
    convert(Ptr{S}, unsafe_convert(Ptr{T}, a))
```

So `Vector{Dual}` → `Ptr{Cdouble}` succeeds because `Dual` is a bits type; C then reads Dual memory as doubles. The same happens for any mismatched bits eltype (e.g. `Vector{Int64}` as `Ptr{Int32}`, homemade dual-like structs, etc.). Base cannot special-case `ForwardDiff.Dual` — detection at the `ccall` boundary is not realistic without a broader `Array{T} → Ptr{S}` safety change (which would be breaking).

### Layering

| Layer | What can be done |
| --- | --- |
| **Base** | Fix docs (#29850); optional warn/check on eltype mismatch (#60268) |
| **SciML** | Document the failure mode + recommend guards (`AutoFiniteDiff`, analytical `jac`, eltype reject before `ccall`, `FunctionWrapperSpecialize`) |

SciML-side FAQ: [DiffEqDocs.jl#880](https://github.com/SciML/DiffEqDocs.jl/pull/880) (merged).

This issue should stay open as a **user-facing DiffEq trap** until the FAQ is enough (and ideally until Base clarifies/docs the behavior). Closing as “Base’s problem only” would lose the autodiff + stiff-solver context that made it hard to debug here.
